csharp/tests/ParallelSelectExtTest.cs

```diff
@@ -0,0 +1,437 @@
+                       : GenerateInts(n)
+                         .ParallelSelect(new ParallelTaskOptions(parallelism),
+                                         f);
+    var x = await enumerable.ToListAsync()
+                            .ConfigureAwait(false);
+    var y = GenerateInts(n)
+      .ToList();
+    Assert.That(x,
+                Is.EqualTo(y));
+  }
+
+  private static IEnumerable ParallelSelectLimitCases()
+  {
+    TestCaseData Case(bool useAsync,
+                      int  parallelism,
+                      int  size)
+      => new TestCaseData(useAsync,
+                          parallelism,
+                          size).SetArgDisplayNames($"int[{size}], {parallelism}{(useAsync ? ", async" : "")}");
+
+    foreach (var useAsync in Booleans)
+    {
+      yield return Case(useAsync,
+                        0,
+                        100);
+      yield return Case(useAsync,
+                        1,
+                        10);
+      yield return Case(useAsync,
+                        2,
+                        20);
+      yield return Case(useAsync,
+                        10,
+                        100);
+      yield return Case(useAsync,
+                        100,
+                        1000);
+      yield return Case(useAsync,
+                        -1,
+                        10000);
+    }
+  }
+
+  [Test]
+  [TestCaseSource(nameof(ParallelSelectLimitCases))]
+  public async Task ParallelSelectLimitShouldSucceed(bool useAsync,
+                                                     int  parallelism,
+                                                     int  n)
+  {
+    var counter    = 0;
+    var maxCounter = 0;
+
+    // We use a larger wait for infinite parallelism to ensure we can actually spawn thousands of tasks in parallel
+    var identity = parallelism < 0
+                     ? AsyncIdentity(200,
+                                     400)
+                     : AsyncIdentity(50,
+                                     100);
+
+    async Task<int> F(int x)
+    {
+      var count = Interlocked.Increment(ref counter);
+      InterlockedMax(ref maxCounter,
+                     count);
+
+      await identity(x)
+        .ConfigureAwait(false);
+      Interlocked.Decrement(ref counter);
+      return x;
+    }
+
+    var enumerable = useAsync
+                       ? GenerateIntsAsync(n)
+                         .ParallelSelect(new ParallelTaskOptions(parallelism),
+                                         F)
+                       : GenerateInts(n)
+                         .ParallelSelect(new ParallelTaskOptions(parallelism),
+                                         F);
+
+    var x = await enumerable.ToListAsync()
+                            .ConfigureAwait(false);
+    var y = GenerateInts(n)
+      .ToList();
+    Assert.That(x,
+                Is.EqualTo(y));
+
+    switch (parallelism)
+    {
+      case > 0:
+        Assert.That(maxCounter,
```

Comment:
I don't really like this way of writing Assert but it seems that it is recommended by a Nunit Analyser. I don't understand why we do not have the analyser warning in Core.

Reply:
I don't know either. One positive with this is that interface makes it clear what is the tested value and what is the expected value.